# **1 - Carregamento dos dados tratados**

In [1]:
import pandas as pd
df = pd.read_csv('dados_tratados_num.csv')
df.head()

Status_Cliente  Sexo  Aposentado  Casado  Dependentes  Meses_Contrato  \
0               0     0           0       0            0               9   
1               0     1           0       1            1               9   
2               1     1           0       1            1               4   
3               1     1           1       0            1              13   
4               1     0           1       0            1               3   

   Assinatura_Telefone  Assinatura_MultiTelefone  Assinatura_Internet  \
0                    0                         0                    0   
1                    0                         1                    0   
2                    0                         0                    1   
3                    0                         0                    1   
4                    0                         0                    1   

   Assinatura_Seguro_online  ...  Assinatura_Seguro_dispositivo  \
0                         0  ...                              0   
1                         0  ...                              0   
2                         0  ...                              1   
3                         0  ...                              1   
4                         0  ...                              0   

   Assinatura_Suporte_tecnico  Assinatura_TV  Assinatura_Filmes  \
0                           0              0                  0   
1                           1              1                  1   
2                           1              1                  0   
3                           1              0                  1   
4                           0              0                  0   

   Tipo_Contrato  Fatura_Online  Forma_Pagamento  Valor_Diario  Valor_Mensal  \
0              0              0                0      2.186667          65.6   
1              1              1                0      1.996667          59.9   
2              1              0                1      2.463333          73.9   
3              1              0                1      3.266667          98.0   
4              1              0                0      2.796667          83.9   

   Valor_Total  
0       593.30  
1       542.40  
2       280.85  
3      1237.85  
4       267.40  

[5 rows x 21 columns]

# **2 - Pré-processamento dos dados**

## **2.1 - Ajuste no Dataframe**

Para evitar quaisquer confusão na interpretação dos dados, a coluna ***Status_Cliente*** será invertida de forma a indicar que:
- 1 = Cliente ativo
- 0 = Cliente inativo

In [2]:
df.Status_Cliente = df.Status_Cliente.replace({0:1,1:0})
df.head()

Status_Cliente  Sexo  Aposentado  Casado  Dependentes  Meses_Contrato  \
0               1     0           0       0            0               9   
1               1     1           0       1            1               9   
2               0     1           0       1            1               4   
3               0     1           1       0            1              13   
4               0     0           1       0            1               3   

   Assinatura_Telefone  Assinatura_MultiTelefone  Assinatura_Internet  \
0                    0                         0                    0   
1                    0                         1                    0   
2                    0                         0                    1   
3                    0                         0                    1   
4                    0                         0                    1   

   Assinatura_Seguro_online  ...  Assinatura_Seguro_dispositivo  \
0                         0  ...                              0   
1                         0  ...                              0   
2                         0  ...                              1   
3                         0  ...                              1   
4                         0  ...                              0   

   Assinatura_Suporte_tecnico  Assinatura_TV  Assinatura_Filmes  \
0                           0              0                  0   
1                           1              1                  1   
2                           1              1                  0   
3                           1              0                  1   
4                           0              0                  0   

   Tipo_Contrato  Fatura_Online  Forma_Pagamento  Valor_Diario  Valor_Mensal  \
0              0              0                0      2.186667          65.6   
1              1              1                0      1.996667          59.9   
2              1              0                1      2.463333          73.9   
3              1              0                1      3.266667          98.0   
4              1              0                0      2.796667          83.9   

   Valor_Total  
0       593.30  
1       542.40  
2       280.85  
3      1237.85  
4       267.40  

[5 rows x 21 columns]

## **2.2 - Eliminação das variáveis não úteis**

Pelo o que foi observado do arquivo de análise gráfica (*Challenge_DS_Alura_Semana02.ipynb*), é possível ter uma noção de quais variáveis não apresentam influência significativa. Assim, a lista de variáveis que não possuem influência na variável ***Status_Cliente*** são:

1. ***Sexo***: O gênero não apresenta influência alguma com a variável de estudo.
2. ***Assinatura_Telefone***: O serviço de assinatura de telefone não apresenta influência na taxa de cancelamento dos clientes. 
3. ***Assinatura_MultiTelefone***: De forma semelhante, o serviço de assinatura multi linha também não paresenta influência na taxa de cancelamento dos clientes.

In [3]:
df.drop(['Sexo','Assinatura_Telefone','Assinatura_MultiTelefone'], axis=1, inplace=True)
df.head()

Status_Cliente  Aposentado  Casado  Dependentes  Meses_Contrato  \
0               1           0       0            0               9   
1               1           0       1            1               9   
2               0           0       1            1               4   
3               0           1       0            1              13   
4               0           1       0            1               3   

   Assinatura_Internet  Assinatura_Seguro_online  Assinatura_Backup  \
0                    0                         0                  0   
1                    0                         0                  1   
2                    1                         0                  1   
3                    1                         0                  0   
4                    1                         0                  1   

   Assinatura_Seguro_dispositivo  Assinatura_Suporte_tecnico  Assinatura_TV  \
0                              0                           0              0   
1                              0                           1              1   
2                              1                           1              1   
3                              1                           1              0   
4                              0                           0              0   

   Assinatura_Filmes  Tipo_Contrato  Fatura_Online  Forma_Pagamento  \
0                  0              0              0                0   
1                  1              1              1                0   
2                  0              1              0                1   
3                  1              1              0                1   
4                  0              1              0                0   

   Valor_Diario  Valor_Mensal  Valor_Total  
0      2.186667          65.6       593.30  
1      1.996667          59.9       542.40  
2      2.463333          73.9       280.85  
3      3.266667          98.0      1237.85  
4      2.796667          83.9       267.40

## **2.3 - Identificação e eliminação das variáveis altamente correlacionadas**

Variáveis altamente correlacionadas podem influenciar negativamente na eficiência no modelo por torna-lo enviesado. Para isto, a tabela de correlação é útil para verificar quais variáveis tem alta correlação (negativa ou positiva) entre si. Assim, observa-se que:

1. ***Valor_Total***: Esta variável possui alta correlação com a variável ***Valor_Mensal***. 
2. ***Valor_Diario***: Esta variável também possui alta correlação com a variável ***Valor_Mensal***. 
3. ***Assinatura_Backup***: Esta variável possui alta correlação com a variável ***Assinatura_Suporte_Tecnico***.
4. ***Assinatura_Seguro_Dispositivo***: Esta variável possui alta correlação com a variável ***Assinatura_Seguro_Online***.

Assim, é possível reduzir estas variáveis em 3 sem perda significativa de informação para o modelo. A saber: ***Valor_Mensal***, ***Assinatura_Suporte*** e ***Assinatura_Segurança***.

In [4]:
# Excluindo as variável Valor_Total, Valor_Diario, Assinatura_Backup, Assinatura_Seguro_Dispositivo
df.drop(['Valor_Total', 'Valor_Diario', 'Assinatura_Backup', 'Assinatura_Seguro_dispositivo'], axis=1, inplace=True)
df.head()

Status_Cliente  Aposentado  Casado  Dependentes  Meses_Contrato  \
0               1           0       0            0               9   
1               1           0       1            1               9   
2               0           0       1            1               4   
3               0           1       0            1              13   
4               0           1       0            1               3   

   Assinatura_Internet  Assinatura_Seguro_online  Assinatura_Suporte_tecnico  \
0                    0                         0                           0   
1                    0                         0                           1   
2                    1                         0                           1   
3                    1                         0                           1   
4                    1                         0                           0   

   Assinatura_TV  Assinatura_Filmes  Tipo_Contrato  Fatura_Online  \
0              0                  0              0              0   
1              1                  1              1              1   
2              1                  0              1              0   
3              0                  1              1              0   
4              0                  0              1              0   

   Forma_Pagamento  Valor_Mensal  
0                0          65.6  
1                0          59.9  
2                1          73.9  
3                1          98.0  
4                0          83.9

In [5]:
# Renomeando as as variáveis correlacionadas
trad = {'Assinatura_Suporte_tecnico':'Assinatura_Suporte', 'Assinatura_Seguro_online':'Assinatura_Seguro'}
df.rename( columns=trad, inplace=True )
df.head()

Status_Cliente  Aposentado  Casado  Dependentes  Meses_Contrato  \
0               1           0       0            0               9   
1               1           0       1            1               9   
2               0           0       1            1               4   
3               0           1       0            1              13   
4               0           1       0            1               3   

   Assinatura_Internet  Assinatura_Seguro  Assinatura_Suporte  Assinatura_TV  \
0                    0                  0                   0              0   
1                    0                  0                   1              1   
2                    1                  0                   1              1   
3                    1                  0                   1              0   
4                    1                  0                   0              0   

   Assinatura_Filmes  Tipo_Contrato  Fatura_Online  Forma_Pagamento  \
0                  0              0              0                0   
1                  1              1              1                0   
2                  0              1              0                1   
3                  1              1              0                1   
4                  0              1              0                0   

   Valor_Mensal  
0          65.6  
1          59.9  
2          73.9  
3          98.0  
4          83.9

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7032 entries, 0 to 7031
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Status_Cliente       7032 non-null   int64  
 1   Aposentado           7032 non-null   int64  
 2   Casado               7032 non-null   int64  
 3   Dependentes          7032 non-null   int64  
 4   Meses_Contrato       7032 non-null   int64  
 5   Assinatura_Internet  7032 non-null   int64  
 6   Assinatura_Seguro    7032 non-null   int64  
 7   Assinatura_Suporte   7032 non-null   int64  
 8   Assinatura_TV        7032 non-null   int64  
 9   Assinatura_Filmes    7032 non-null   int64  
 10  Tipo_Contrato        7032 non-null   int64  
 11  Fatura_Online        7032 non-null   int64  
 12  Forma_Pagamento      7032 non-null   int64  
 13  Valor_Mensal         7032 non-null   float64
dtypes: float64(1), int64(13)
memory usage: 769.2 KB


# **3. Análise do balanceamento da variável de estudo *Status_Cliente***

## **3.1 - Proporção de clientes ativos e inativos**

Relembrando a análise quantitativa do relatório anterior, obtém-se a seguinte proporção de clientes ativos e inativos:

In [7]:
print('----------------------------------')
print('Quantidade de clientes ativos (1) e inativos (0)')
print(df.Status_Cliente.value_counts())
print('----------------------------------')
print('Quantidade percentual de clientes ativos (1) e inativos (0)')
print(df.Status_Cliente.value_counts(normalize=True).map("{:.1%}".format))
print('----------------------------------')

----------------------------------
Quantidade de clientes ativos (1) e inativos (0)
1    5163
0    1869
Name: Status_Cliente, dtype: int64
----------------------------------
Quantidade percentual de clientes ativos (1) e inativos (0)
1    73.4%
0    26.6%
Name: Status_Cliente, dtype: object
----------------------------------


## **3.2 - Técnica de balanceamento para distribuições desbalanceadas**

Como a variável ***Statuts_Cliente*** tem a proporção de 1:4 das respostas (inativo para ativo), é aconselhável realizar uma verificação das técnicas de balanceamento. Por mais que a proporção da classe de registro minoritária (Clientes inativos) seja relativamente grande, em torno de 26.5%, o modelo ainda pode sofrer certo grau de enviesamento. Para isto, algumas técnicas de balanceamento serão analisadas. Tais técnicas consistem em reduzir a disparidade entre a classe marjoritária (neste caso, cliente ativo) e a minoritária (cliente inativo). Além disto, também é possível realizar tratamento de registros que possam ser considerados ambíguos pelos algoritmos de classificação. Para este projeto, serão utilizados as seguintes técnicas:

- **Randon Under Sampler**: A técnica de desbalaceamento RandomSampling é a mais simples e consiste na escolha aleatória de registros da classe marjoritária para o descarte, de forma a aumentar a diminuir a disparidade entre as classes minoritárias e marjoritárias.

- **NearMiss**: Near Miss é um conjunto de técnicas UnderSampling the seleciona os exemplos baseado na distância dos dados marjoritários.

- **Condesed Nearest Neighbour**: O método Condensed Nearest Neighbour (CNN) é uma técnica de de UnderSampling que procura o conjunto de amostras de uma classe que resulta em perda mínima da performance do modelo, referenciado como conjunto minimamente consistente. 

- **One-Sided Selection**: O método One-Sided Selection (OSS) é uma combinação dos métodos Tomek Link e Condensed Nearest Neighbor (CNN). O método OSS então  utiliza o Tomek Link para remover os exemplos ambíguos e depois utiliza o método CNN para remover os exemplos redundantes que estão distantes dos contornos das classes.

- **Neighborhood Cleaning Rule**: O método Neighborhood Cleaning Rule (NCR) consiste na combinação do método ENS para o tratamento de exemplos ambíguos e o método CNN para o tratamento dos exemplos redundantes. Entretanto, diferentemente do método OSS, o método CNR esta mais focado na limpeza dos exemplos que necessariamente na remoção de exemplos redundantes. Portanto, o método NCR está mais focado no tratamento dos exemplos ambíguos.   

In [8]:
# Separando as variáveis
y = df.Status_Cliente
X = df.drop(['Status_Cliente'], axis=1)

In [9]:
def under_sampling(type, seed_state, x, y):
    # Avaliação de qual técnica será utilizada
        # RUS -> Random Under Sampler
        # NM  -> Near Miss
        # CNN -> Condesed Nearest Neighbour
        # OSS -> One-Sided Selection
        # NCR -> Neighborhood Cleaning Rule
        # None -> Nenhuma técnica será aplicada

    # Utilizando o Randon Under Sampler
    if type == 'RUS':
        from imblearn.under_sampling import RandomUnderSampler

        undersample = RandomUnderSampler(random_state=seed_state)

        x_resampled, y_resampled = undersample.fit_resample(x,y)
    
    # Utilizando o NearMiss
    elif type == 'NM':
        from imblearn.under_sampling import NearMiss

        versao=1
        n=3
        undersample = NearMiss(n_neighbors=n,version=versao)

        x_resampled, y_resampled = undersample.fit_resample(x,y)

    # Utilizando o Condesed Nearest Neighbour
    elif type == 'CNN':
        from imblearn.under_sampling import CondensedNearestNeighbour

        undersample = CondensedNearestNeighbour(n_neighbors=1)

        x_resampled, y_resampled = undersample.fit_resample(x,y)
    
    # Utilizando o One-Sided Selection
    elif type == 'OSS':
        from imblearn.under_sampling import OneSidedSelection

        undersample = OneSidedSelection(n_neighbors=1, n_seeds_S=200)

        x_resampled, y_resampled = undersample.fit_resample(x,y)
    
    # Utilizando o Neighbourhood Clean Rule
    elif type == 'NCR':
        from imblearn.under_sampling import NeighbourhoodCleaningRule

        undersample = NeighbourhoodCleaningRule(n_neighbors=3, threshold_cleaning=0.5)

        x_resampled, y_resampled = undersample.fit_resample(x,y)
    
    # Sem nenhum tratamento de balanceamento
    else:
        x_resampled = x
        y_resampled = y

    return x_resampled, y_resampled


# **4 - Avaliação dos modelos de Machine Learning**

Como etapa preliminar, uma função será construída compilando os principais algoritmos de Machine Learning para classificação. Para isto, serão utilizados os seguintes algoritmos de classficação supervisionados:

- Logistic Regression:

    O Logistic Regression utiliza a função de Sigmoid para gerar a probabilidade de uma determinada classificação. Este algoritmo é largamente utilizado quando o dataset possui uma classificação binária (0/1, verdadeiro/falso, etc). 

    Para o caso de um dataset multiclasse, o algoritmo utiliza a abordagem um versus restante. É possível configurar isto definindo o parâmetro multi_class como ovr.  Também é possível ajustar o solver (solver), penalidade (penalty) e a distribuição da classe (class_weight). 

- Decision Tree Classifier:

    O algoritmo Desicion Tree Classifier é um algoritmo de classificação onde as características mais importantes da dos dados vão sendo subdivididos em avaliações de verdadeiro ou falso. A partição dos dados permite montar um padrão de classificação por meio das ramificação da árvore, onde um determinado registro pode ser classificado ao ser comparado com a árvore de decisão.

- Random Forest:

    O algoritmo Random Forest é uma combinação de vários Decisions Tree. Este algoritmo utiliza a técnica de sub-amostragem e aplica a média das amostras para melhorar a previsão e contralar o problema de overfiting. 

- k-Neighbor Classifier:

    Este algoritmo utiliza a abordagem de distância entre os registro em um espeço vetorial correspondente. Registros que possuem uma distância média pequena tendem a ter a mesma classificação. Este tipo de algoritmo também é bastante utiliado como sistema de recomendação. 

- Gaussian Nayve-Bayes:

    O algoritmo Gaussian Navye-Bayes é baseado no Teorema de Bayes onde a probabilidade condicional é calculada com base em uma informação dada. A condição de Naive assume que as informações dos registros são independentes (outro motivo para a eliminação das variáveis altamente correlacionadas). A grande vantagem deste algoritmo é que, diferentemente dos demais algoritmos de aprendizado, este não necessita de um grande banco de dados para apresentar uma boa performance nas previsões.

- Support Vector Machine:

    O Support Vector Machine é um algoritmo onde utiliza de um hiper palno para separar os registros em duas classes: negativa e positiva. Este hiper plano é calculado com base na maximização da distância entre os pontos. Com o hiper plano definindo a borda entre as duas classes, o algoritmo realiza a previsão da classificação com base na posição de um determinado registro neste espaço vetorial.

    Este algortimo, assim como Decision Tree e o Random Forest, pode ser utilizado tanto para classificação quanto para regresão.

- Ada Boost:

    O algoritmo Ada Boost é nada mais que um compilado de vários algoritmos de previsão mais simples (e menos eficientes) organizados de forma iterativa de modo a fornecer, no conjunto, um algoritmo de previsão com alta eficiência. A ideia é avaliar uma combinação linear entre os resultados entre vários sub-modelos e otimizar os pesos de tais modelos de forma a fornecer a maior acurácia possível.

- Gradient Boosting Classifier:

    De forma semelhante ao Ada Boost, o Gradient Boosting Classifier configura-se como um compilado de árvores de decisão (Decission Trees). 

## **4.1 - Função para avaliar os modelos**

In [10]:
# Criando a função para avaliar diferentes algoritmos
from sklearn.model_selection import cross_validate 

def classifier_model(model,x,y,n_folders):

    # Resultados da cross-validação
    results_cv = cross_validate(model, x, y, cv=n_folders, scoring = ['precision','recall', 'accuracy'])
    mean_accuracy = round( results_cv['test_accuracy'].mean(), 3 )
    mean_precision = round( results_cv['test_precision'].mean(), 3 )
    mean_recall = round( results_cv['test_recall'].mean(), 3 )

    # Retornanco os dados
    return [mean_accuracy, mean_precision, mean_recall]

## **4.2 - Modelos de classificação**

In [15]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier

# Definindo os modelos a serem utilizados
seed_state = 101 # Fizando o seed do random generator
lr = LogisticRegression(random_state=seed_state, max_iter = 500)
dtc = DecisionTreeClassifier(random_state=seed_state)
rfc = RandomForestClassifier(random_state=seed_state)
knc = KNeighborsClassifier()
gnb = GaussianNB()
svc = SVC(random_state=seed_state)
abc = AdaBoostClassifier(random_state=seed_state, learning_rate = 1.5)
gbc = GradientBoostingClassifier(random_state=seed_state)

# Compilando os mdoelos
models = [lr, dtc, rfc, knc, gnb, svc, abc, gbc]
models_name = ['LogisticRegression', 'DecisionTreeClassifier', 'RandomForestClassifier', 'KNeighborsClassifier', 'GaussianNB', 
    'SupportVectorMachine', 'AdaBoostClassifier','GradientBoostingClassifier' ]

## **4.3 - Compração entre os modelos**

Para a comparação dos modelos, a técnica de balanceamento utilizada foi a NCR devido à sua robustez. Fixando o método de balanceamento, observou-se que o modelo *K-Neighbors Classifier* apresentou a maior precisão e acurácia entre os modelos avaliados. 

In [16]:
# Parâmetros do balanceamento 
type = 'NCR'     # Definindo a técnica de balanceamento
seed_state = 101
cv_number = 10

# Resultado dos modelos
models_results = []

print('-------------------------------------------')
print('Calculando...')
for i, model in enumerate(models):
    print('     ',models_name[i])

    x_resample, y_resample = under_sampling(type, seed_state, X, y)
    result = classifier_model(model, x=x_resample, y=y_resample, n_folders=cv_number)
    models_results.append(result)

print('Cálculos finalizados!')
print('-------------------------------------------')
models_results = pd.DataFrame(models_results, index=models_name, columns=['Accuracy', 'Precision','Recall'])
models_results.sort_values(by=['Accuracy', 'Precision', 'Recall'], ascending=False)

-------------------------------------------
Calculando...
      LogisticRegression
      DecisionTreeClassifier
      RandomForestClassifier
      KNeighborsClassifier
      GaussianNB
      SupportVectorMachine
      AdaBoostClassifier
      GradientBoostingClassifier
Cálculos finalizados!
-------------------------------------------


Accuracy  Precision  Recall
KNeighborsClassifier           0.861      0.882   0.918
RandomForestClassifier         0.861      0.877   0.924
GradientBoostingClassifier     0.855      0.875   0.918
AdaBoostClassifier             0.846      0.878   0.898
LogisticRegression             0.841      0.865   0.907
SupportVectorMachine           0.818      0.827   0.926
DecisionTreeClassifier         0.809      0.864   0.853
GaussianNB                     0.803      0.903   0.796

## **4.4 - Otimizando o modelo**

Para a otimização do modelo, foi utilizada a função *RandomizedSearchCV* para encontrar os melhores parâmetros do modelo *K-Neighbors Classifier*. 

In [22]:
from sklearn.model_selection import train_test_split
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint
from sklearn.model_selection import KFold

In [26]:
grid_parameters = {
    'n_neighbors' : randint(3,20),
    'weights' : ['uniform', 'distance'],
    'metric' : ['euclidean', 'manhattan']
}

grid_s = RandomizedSearchCV( knc, grid_parameters, n_iter=20, 
                            cv= KFold(n_splits = 5, random_state =seed_state, shuffle=True), 
                            n_jobs=-1 )

X_rs, y_rs = under_sampling('NCR', seed_state, X,y)
X_train, X_test, y_train, y_test = train_test_split(X_rs, y_rs, test_size=0.25, random_state=seed_state)
grid_s.fit(X_train,y_train)
results_opt = pd.DataFrame(grid_s.cv_results_)
results_opt = results_opt.filter( ['rank_test_score','mean_test_score','std_test_score',
    'mean_fit_time','std_fit_time','params'])
results_opt.sort_values(by='rank_test_score',ascending=True).head(3)

rank_test_score  mean_test_score  std_test_score  mean_fit_time  \
19                1         0.864902        0.010746       0.023797   
15                2         0.864445        0.010171       0.026799   
0                 2         0.864445        0.010171       0.021795   

    std_fit_time                                             params  
19      0.002484  {'metric': 'manhattan', 'n_neighbors': 11, 'we...  
15      0.005078  {'metric': 'manhattan', 'n_neighbors': 12, 'we...  
0       0.003063  {'metric': 'manhattan', 'n_neighbors': 12, 'we...

# **5 - Conclusões**

Esta última etapa teve por objetivo construir modelos preditivos para avaliar o perfil de clientes da empresa AluraVoz. Para a utilização do dataset nos modelos de classificação, algumas variáveis foram eliminadas e/ou alteradas para evitar redundância nos dados de treino e teste. 

Como a proporção de cliente inativos e ativos é de 4:1, algumas técnicas de balanceamento foram explorados ao longo deste relatório. Como a proporção é razoavelmente equilibrada, o método NCR, onde somente os dados redundantes são eliminados, apresentou melhor performance. Quanto ao modelos, foram avaliados diversos algoritmos e o *k-NeighborsClassifier* foi o que apresentou maior acurácia (86.1%). Por fim, o processo de otimização hiperparamétrica foi avaliado e o melhor conjunto de parâmetros encontrado fornecer uma precisão média de 86.5%.